### DISCUSSION TITLE
Add UUID5 support in Morphir.

### DISCUSSION THREAD

**Discussion by DamianReeves** [finos/morphir#231] [Proposal]
UUID5 is a deterministic version of the UUID. 
UUID's are used in business logic to represent IDs, account numbers, keys, and other similar information. 
While the initial inclination may be that this type has no place in business logic, there are in fact business cases which we have that would benefit from supporting such a type.

In the case of card numbers, we often don't want the actual card number to land in logs and other systems which can capture the card number or account number. UUID v3 & v5 both provide a deterministic hash based manner for creating a UUID which accomplishes this but would still allow us to use the meaningful account/card number in calculations.

These calculations, and logic may even include the actual production of virtual card numbers (which are just "virtual" representatiosn of a card).

**Comment by stephengoldbaum** [comment]
This fits into the category of structured string types.  Bosque has [Typed Strings](https://github.com/microsoft/BosqueLanguage) for this purpose.  How do you propose adding these to Morphir?

**Reply by DamianReeves** [reply]
I suggest we directly add a type for UUID.
I see your note on structured strings and it also makes me think of TypeSpec's [scalars](https://typespec.io/docs/language-basics/scalars). 

In the case of UUID, I suggest a UUID module for UUID, but I also suggest to support "Scalars" we add a Scalar module to support Scalar types or perhaps a sort of annotations module as a way to add custom annotations/decorations which are supported directly by Morphir. It reminds me a bit how the [Fable Compiler](https://fable.io/docs/javascript/features.html) and other Language to JS compilers allow you to provide metadata that may be "erased" at runtime.

**Reply by stephengoldbaum** [reply]
What's the advantage of adding as a type?  What trade-offs should we consider?  I can imagine there being many of these (UUID, URL, IRI, CUSIP, ISIN, etc.) and this approach becoming unwieldy.

**Reply by DamianReeves** [reply]
URI and UUID are foundational as opposed to CUSIP, ISIN etc.
Without support we can't properly build URI (you could build it as a struct, but not parse it from a string) and UUID.
I'm not saying adding all things like CUSIP and ISIN, but add the foundational types like these that are needed to build the other more business specific types.

**Comment by stephengoldbaum** [comment]
Can you explain what "support" means specifically and how it helps with
parsing it from a String?
